core/controllers/acl_decorators.py

```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
         Raises:
             PageNotFoundException: The given page cannot be found.
         """
-        topic_name = python_utils.url_unquote_plus(topic_name)
-        topic = topic_fetchers.get_topic_by_name(topic_name)
+        lowercase_abbrev_topic_name = (
+            python_utils.url_unquote_plus(abbrev_topic_name).replace('-', ' '))
+        topic = topic_fetchers.get_topic_by_abbreviated_name(
+            lowercase_abbrev_topic_name)
+        classroom_name = kwargs.pop('classroom_name', None)
 
         if topic is None:
-            raise self.PageNotFoundException
+            handle_redirection(
+                self, '/%s' % classroom_name,
+                self.GET_HANDLER_ERROR_RETURN_TYPE)
+            return
+        if classroom_name not in [None, constants.CLASSROOM_STATUS_STAGING]:
+            lowercase_classroom_name = (
+                python_utils.url_unquote_plus(
```

Comment:
Please try and design the fields etc. so as to remove the need for this construction anywhere. It is preferable to avoid string manipulation where possible if it isn't necessary.

Reply:
Then we need to add such a field for classrooms.